scipy/sparse/linalg/matfuncs.py

```diff
@@ -70,6 +70,10 @@ def inv(A):
     .. versionadded:: 0.12.0
 
     """
+    #check input
+    if not scipy.sparse.isspmatrix(A):
+        raise TypeError('expected A to be a sparse matrix')
```

Comment:
It is a wrong supplied argument so ValueError needs to be raised.

Reply:
This is not something I'm very knowledgeable about... but the docs for type error say

"Raised when an operation or function is applied to an object of inappropriate type. The associated value is a string giving details about the type mismatch."

That seems correct here: `inverse` is passed a type that isn't a sparse matrix. Thoughts?

Reply:
This is indeed something sitting on the bench. When viewed from a function perspective it's just a wrong argument. When viewed from the array it is a TypeError raising operation. Come to think of it, I don't have any strong opinions about it though my initial comment sounded otherwise. I usually reserve this exception to up/downcasting and alike. 

Pandas people also called it a gray area https://github.com/pandas-dev/pandas/wiki/Choosing-Exceptions-to-Raise hence 

I'm OK with it if that is the consensus.